docs/dg/dev/architecture/coding-conventions.md

```diff
@@ -0,0 +1,2165 @@
+
+        return $this->jsonResponse($table->fetchData());
+    }
+}
+```
+
+## Dependency Provider
+
+```
+[Organization]
+├── Zed
+│   └── [Module]        
+│       ├── [Module]Config.php
+│       └── [Module]DependencyProvider.php
+├── Yves
+│   └── [Module]        
+│       ├── [Module]Config.php
+│       └── [Module]DependencyProvider.php
+├── Glue
+│   └── [Module]        
+│       ├── [Module]Config.php
+│       └── [Module]DependencyProvider.php
+├── Client
+│   └── [Module]        
+│       ├── [Module]Config.php
+│       └── [Module]DependencyProvider.php
+└── Service
+    └── [Module]        
+        ├── [Module]Config.php
+        └── [Module]DependencyProvider.php
+```
+
+**Description**
+
+Injects required dependencies to a module application layer. Typically, dependencies are [facades](#facade-design-pattern) or [plugins](#plugin). 
+
+Dependency injection is orchestrated through a `provide-add-get` structure (see example below).
+- The `provide` method is the highest level that holds only `add` calls without any additional logic.
+- The `add` method is the middle level that injects the dependencies into the dependency container using a class constant and a late-binding instantiating closure.
+- The `get` method is the lowest level that sources the dependency.
+
+**Conventions**
+- All class constants MUST be `public` (to decrease conflicts in definition for being a public API class).
+- Each dependency MUST be defined using a class constant.
+- Setting dependency using the `container::set()` MUST be paired with late-binding closure definition to decouple instantiation.
+- Dependencies MUST be wired through the target layer corresponding inherited method:
+```php
+public function provideCommunicationLayerDependencies(Container $container)
+public function provideBusinessLayerDependencies(Container $container)
+public function providePersistenceLayerDependencies(Container $container)
+public function provideDependencies(Container $container)
+public function provideBackendDependencies(Container $container)
+public function provideServiceLayerDependencies(Container $container)
+```
+- Only three type of methods CAN be defined, either `provide`, `get`, or `add`.
+```php
+public function provide*Dependencies(Container $container)
+```
+- `Provide` methods MUST call their parent `provide` method.
+- `Provide` methods MUST only call `add` methods.
+
+```php
+protected function add*(Container $container)
+```
+- `Add` methods MUST only introduce one dependency to the `Container` (a plugin-stack is considered one dependency in this respect).
+
+```php
+protected function get[Dependency](Container $container)
+protected function get[PluginInterfaceName]Plugins(Container $container)
+```
+
+<details><summary markdown='span'>Additional Conventions for Project Development</summary>
+- Constants can have any access modifier, they are not limited to `public`. <br/>
```

Comment:
Having protected and private access level does not bring any value. Is there any use-case where non-public access level will be useful?
```suggestion
```

Reply:
yes, there are.
For example, we reuse a plugin stack from Calculate::quote in the Calculation::order, by unsetting/replacing some of the plugins by keys, and the keys are private/protected, since used only in this dependency provider.

Why projects should not be allowed to add some protected constant in this class?

Reply:
Because in my mind, the class constants were supposed to be used in both dependency providers and factories. So, if you made them private/protected, they would be invisible to factories.

The use-case you shared is interesting one. I would group common plugins for quote and order inside of a method, and add new ones without the need of replacing, or unsetting.

In any-case this is a valid use case :). I will close my comment.